### ISSUE TITLE
Swift 6 concurrency error on `_StorageClass.defaultInstance`

### ISSUE THREAD

**Issue by rebello95** [apple/swift-protobuf#1729]
When building with Xcode 16.0 and the Swift 6 toolchain, the following error occurs with a generated `_StorageClass` type:

> Static property 'defaultInstance' is not concurrency-safe because non-'Sendable' type

<img width="1135" alt="Screenshot 2024-10-03 at 10 42 56 AM" src="https://github.com/user-attachments/assets/2171b604-ec71-44c3-89b4-8a62a91533a0">

I read through https://github.com/apple/swift-protobuf/issues/1560 which discussed this a bit, but based on [this comment](https://github.com/apple/swift-protobuf/issues/1560#issuecomment-2364523165) I opted to open a new issue.

This can reproduced on [this Connect-Swift branch](https://github.com/connectrpc/connect-swift/commit/403398d43f8cd78c813ba42e283f7e54ac03fb3a), but I think this is an issue with any generated `_StorageClass` when building with Swift 6.

**Comment by thomasvl** [comment]
@FranzBusch @gjcairo fyi - since I think you all mentioned working on full Swift 6 language mode.  I guess starting with generated code would unblock folks first.

While #1561 might be sorta large to pull off in a non breaking way, I do wonder if that resolves a lot of this since I think the name map could get a lot simpler (no need for interning, etc.)

**Comment by thomasvl** [comment]
@tbkka fyi also

**Comment by FranzBusch** [comment]
> since I think you all mentioned working on full Swift 6 language mode. I guess starting with generated code would unblock folks first.

I agree tackling the generated code is the most important thing. I do think we need to do all of this in a non-breaking way for now. Right now we are working through our other repos to make sure they compile without errors in Swift 6.

**Comment by Lukasa** [comment]
I think it'd be very helpful to get a complete list of Swift 6 issues you see. The specific concurrency issue here should be solveable using `nonisolated(unsafe)` on this `static let`, but I'd love to know if there were more.

**Comment by thomasvl** [comment]
Can someone that is seeing this issue see if they can recreated it by modifying https://github.com/apple/swift-protobuf/pull/1737?  I made attempt at ensuring the generate file had the _Storage_ and also enabling v6 language mode on the packages, but I can't seem to get the warnings/errors in the new CI step.

**Comment by rebello95** [comment]
> I think it'd be very helpful to get a complete list of Swift 6 issues you see. The specific concurrency issue here should be solveable using nonisolated(unsafe) on this static let, but I'd love to know if there were more.

@Lukasa Manually changing that line to `nonisolated(unsafe) static let defaultInstance = _StorageClass()` allowed me to build without further errors/warnings in my case.

> Can someone that is seeing this issue see if they can recreated it by modifying https://github.com/apple/swift-protobuf/pull/1737? I made attempt at ensuring the generate file had the Storage and also enabling v6 language mode on the packages, but I can't seem to get the warnings/errors in the new CI step.

@thomasvl I wonder if this could be related to the fact that CI is [running](https://github.com/apple/swift-protobuf/blob/43bc712f0152fcca7f8b950d9409bcd8fc20a40e/.github/workflows/build.yml#L14C14-L14C27) on `ubuntu-latest` 🤔 are you able to repro the issue locally?

**Comment by thomasvl** [comment]
> @thomasvl I wonder if this could be related to the fact that CI is [running](https://github.com/apple/swift-protobuf/blob/43bc712f0152fcca7f8b950d9409bcd8fc20a40e/.github/workflows/build.yml#L14C14-L14C27) on `ubuntu-latest` 🤔 are you able to repro the issue locally?

On my Mac with Xcode 16.2 installed, using the github _main_ branch, `make test-spm-plugin SWIFT_BUILD_TEST_HOOK="-Xswiftc -warnings-as-errors"` passes.  So, no; I can't seem to repo it, so I'm hoping other folks that see the failure(s) can provide some insights so we can get something to repo it to track it going forward.

**Comment by thomasvl** [comment]
@rebello95 skimming back through, I don't think it was said, what version of SwiftProtobuf are you pulling in?  Are you maybe on an older release?

**Comment by rebello95** [comment]
> @rebello95 skimming back through, I don't think it was said, what version of SwiftProtobuf are you pulling in? Are you maybe on an older release?

I'm seeing this on 1.28.2. It's reproducible on [this commit](https://github.com/connectrpc/connect-swift/commit/89b03f3a5f9d346be4a80cc36af9ea4d4854f8b4) when trying to build the Connect-Swift project in Xcode 16.2.

**Comment by thomasvl** [comment]
> > @rebello95 skimming back through, I don't think it was said, what version of SwiftProtobuf are you pulling in? Are you maybe on an older release?
> 
> I'm seeing this on 1.28.2. It's reproducible on [this commit](https://github.com/connectrpc/connect-swift/commit/89b03f3a5f9d346be4a80cc36af9ea4d4854f8b4) when trying to build the Connect-Swift project in Xcode 16.2.

That commit is changing the _main_ `Package.swift`, so it is changing how the SwiftProtobuf _runtime_ is compiled, it isn't just about the generated code, and I thought your original comment in opening this issue was just about the generated code.  How are you getting the original error from generated code without changing out the SwiftProtobuf code itself is compiled?  SwiftPM should be fine with this library compiled in Swift 5 Language Mode and your code (and the generated code for Protos) being compiled in the 6 mode. If you have a problem in that setup, we'd like to recreate that to ensure we keep that building at all times.
